File: Tree/Tree_Depth_First/dfs_tree.cpp
#include <iostream>

// Tree Node implementation
struct Node {
    int data;
    struct Node* left, *right;
    Node(int data) {
        this->data = data;
        left = nullptr;
        right = nullptr;
    };
};

class DFSTree {
    private:
        // Root of the tree
        struct Node* root;

        // In order traversal helper function.
        void inOrderHelper(const struct Node*);

        // Pre order traversal helper function.
        void preOrderHelper(const struct Node*);

        // Post order traversal helper function.
        void postOrderHelper(const struct Node*);

    public:
        // Constructor
        DFSTree() { this->root = nullptr; };

        // Insert overload to recur
        struct Node* insert(struct Node*, struct Node*);

        // Inserts a new node in the tree
        void insert(const int);

        /*
            Given a bin tree, prints its nodes according
            to the "bottom-up" inorder traversal.
                (LEFT, ROOT, RIGHT)
        */
        void inOrder(void);

        /*
            Given a bin tree, prints its nodes according
            to the "bottom-up" preorder traversal.
                (ROOT, LEFT, RIGHT)
        */
        void preOrder(void);

        /* 
            Given a bin tree, prints its nodes according
            to the "bottom-up" postorder traversal.
                (LEFT, RIGHT, ROOT)
        */
        void postOrder(void);
};

int main(void) {
    DFSTree tree;

    tree.insert(5); 
    tree.insert(3); 
    tree.insert(2); 
    tree.insert(4); 
    tree.insert(7); 
    tree.insert(6); 
    tree.insert(8);

    std::cout << "Postorder: \n";
    tree.postOrder();

    std::cout << "\nInorder:  \n";
    tree.inOrder();

    std::cout << "\nPreorder:  \n";
    tree.preOrder();

    return 0;
};

void DFSTree::postOrderHelper(const struct Node* node) {
    if (node == nullptr) return;

    // Go left first
    postOrderHelper(node->left);

    // Then go right
    postOrderHelper(node->right);

    // Finally go into node's data
    std::cout << node->data << " ";
};

void DFSTree::inOrderHelper(const struct Node* node) {
    if (node == nullptr) return;

    // First go left
    inOrderHelper(node->left);

    // Then into node's data
    std::cout << node->data << " ";

    // Finally to right child
    inOrderHelper(node->right);
};

void DFSTree::preOrderHelper(const struct Node* node) {
    if (node == nullptr) return;

    // Start with node's data
    std::cout << node->data << " ";

    // Then go left subtree
    preOrderHelper(node->left);

    // Finally right
    preOrderHelper(node->right);
};

void DFSTree::postOrder() {
    postOrderHelper(root);
};

void DFSTree::preOrder() {
    preOrderHelper(root);
};

void DFSTree::inOrder() {
    inOrderHelper(root);
};

void DFSTree::insert(const int data) {
    struct Node* node = new Node(data);
    if (root == nullptr) root = node;

    insert(root, node);
};

struct Node* DFSTree::insert(struct Node* root, struct Node* node) {
    if (root == nullptr) return node;
    
    if (node->data < root->data) {
        root->left = insert(root->left, node);
    } else if (node->data > root->data) {
        root->right = insert(root->right, node);
    }

    return root;
};

File: Tree/Tree_Depth_First/DFSTree.java


/**
 * Depth-First Binary Search Tree implementation class.
 * @author Yasin Radi
 */
public class DFSTree {

    /** Tree Node implementation. */
    private class Node {
        // Data held by the node
        private int data;

        // Left child of node
        private Node left;

        // Right child of node
        private Node right;

        private Node(int data) {
            this.data = data;
            this.right = null;
            this.left = null;
        };
    };

    // Root of the tree
    private Node root;

    public DFSTree() {
        this.root = null;
    };

    /**
     * Inserts a node with the given data into the tree.
     * @param data  int
     */
    public void insert(final int data) {
        // Create node to be inserted
        Node node = new Node(data);

        // Perform a normal BST insert.
        this.root = insertHelper(this.root, node);
    };

    /**
     * Helper that performs a std BST insert.
     * @param root  Node
     * @param node  Node
     * @return      Node
     */
    private Node insertHelper(Node root, Node node) {
        // If tree is empty, return new node
        if (root == null) {
            return node;
        }

        // Otherwise, recur down the tree
        if (node.data < root.data) {
            root.left = insertHelper(root.left, node);
        } else if (node.data > root.data) {
            root.right = insertHelper(root.right, node);
        }

        // Return the unchanged node
        return root;
    };

    /** Helper function to postOrder traversal. */
    private void postOrderHelper(final Node node) {
        if (node == null) {
            return;
        }

        // Incur first in left subtree
        postOrderHelper(node.left);

        // Then go right subtree
        postOrderHelper(node.right);

        // Finally deal with the node itself
        System.out.print(node.data + " ");
    };

    /** Helper function for preOrder traversal. */
    private void preOrderHelper(final Node node) {
        if (node == null) {
            return;
        }

        // Start with the node's data
        System.out.print(node.data + " ");

        // Then go left subtree
        preOrderHelper(node.left);

        // Finally right
        preOrderHelper(node.right);
    };

    /** Helper function for inOrder traversal. */
    private void inOrderHelper(final Node node) {
        if (node == null) {
            return;
        }

        // Incur first in left subtree
        inOrderHelper(node.left);

        // Then print the data in the node
        System.out.print(node.data + " ");

        // Then go right
        inOrderHelper(node.right);
    };

    /**
     * Given a bin tree, prints its nodes according
     * to the "bottom-up" postorder traversal.
     * (LEFT, RIGHT, ROOT)
     */
    public void postOrder() {
        postOrderHelper(this.root);
        System.out.println();
    };

    /**
     * Given a bin tree, prints its nodes according
     * to the "bottom-up" inorder traversal.
     * (LEFT, ROOT, RIGHT)
     */
    public void inOrder() {
        inOrderHelper(this.root);
        System.out.println();
    };

    /**
     * Given a bin tree, prints its nodes according
     * to the "bottom-up" preorder traversal.
     * (ROOT, LEFT, RIGHT)
     */
    public void preOrder() {
        preOrderHelper(this.root);
        System.out.println();
    };
};